vendor/gopkg.in/src-d/go-git.v4/remote.go

```diff
@@ -522,7 +544,36 @@ func getHaves(localRefs []*plumbing.Reference) ([]plumbing.Hash, error) {
 			continue
 		}
 
-		haves[ref.Hash()] = true
+		// No need to load the commit if we know the remote already
+		// has this hash.
+		if stopEarlyRefs[ref.Hash()] {
+			haves[ref.Hash()] = true
+			continue
+		}
+
+		commit, err := object.GetCommit(s, ref.Hash())
+		if err != nil {
+			haves[ref.Hash()] = true
+			continue
+		}
+
+		// Until go-git supports proper commit negotiation during an
+		// upload pack request, include up to 100 commits from the
+		// history of each ref.
+		walker := object.NewCommitPreorderIter(commit, haves, nil)
+		const maxToVisit = 100
```

Comment:
What if there isn't a common commit in the last 100? Does it degrade to transferring all the refs?

Reply:
Yep, exactly.